### PR TITLE
Crash fix for tensor.concatenate with negative axis and fast_run

### LINKED ISSUE #2747: Crash for T.concatenate([x, x], axis=-1) grad
Hi Theano devs, thanks for writing this great tool.

When I use the fast_run optimizer the following code raises an exception. It works when I use fast_compile, or when I replace the axis=-1 with axis=0 (or whatever the last axis of the argument may be). Since this works under one optimizer and not under another, I think its is a bug and not just a missing feature.

```python
import numpy as np
import theano
import theano.tensor as T

theano.config.exception_verbosity = 'high'
theano.config.optimizer = 'fast_run'
#theano.config.optimizer = 'fast_compile'

X = T.dvector()  # or dmatrix, dtensor3, ...
f = T.concatenate([X, X], axis=-1).sum()  # This fails with fast_run but succeeds with fast_compile
#f = T.concatenate([X, X], axis=0).sum()  # This always works
grad = T.grad(f, wrt=[X])
fg = theano.function([X], grad)
fg(np.random.randn(10))
```

The error I get is:
```
python concat_test.py 
Traceback (most recent call last):
  File "concat_test.py", line 15, in <module>
    fg(np.random.randn(10))
  File "/Users/tsc/anaconda/lib/python2.7/site-packages/theano/compile/function_module.py", line 606, in __call__
    storage_map=self.fn.storage_map)
  File "/Users/tsc/anaconda/lib/python2.7/site-packages/theano/compile/function_module.py", line 595, in __call__
    outputs = self.fn()
  File "/Users/tsc/anaconda/lib/python2.7/site-packages/theano/gof/op.py", line 768, in rval
    r = p(n, [x[0] for x in i], o)
  File "/Users/tsc/anaconda/lib/python2.7/site-packages/theano/tensor/basic.py", line 3205, in perform
    (numpy.sum(splits), len_along_axis))
ValueError: The splits sum to 20, expected 10
Apply node that caused the error: Split{2}(Alloc.0, TensorConstant{-1}, MakeVector.0)
Inputs types: [TensorType(float64, vector), TensorType(int8, scalar), TensorType(int64, vector)]
Inputs shapes: [(10,), (), (2,)]
Inputs strides: [(8,), (), (8,)]
Inputs values: ['not shown', array(-1, dtype=int8), array([10, 10])]

Backtrace when the node is created:
  File "/Users/tsc/anaconda/lib/python2.7/site-packages/theano/gradient.py", line 1089, in access_term_cache
    input_grads = node.op.grad(inputs, new_output_grads)

Debugprint of the apply node: 
Split{2}.0 [@A] <TensorType(float64, vector)> ''   
 |Alloc [@B] <TensorType(float64, vector)> ''   
 | |TensorConstant{(1,) of 1.0} [@C] <TensorType(float64, (True,))>
 | |Shape_i{0} [@D] <TensorType(int64, scalar)> ''   
 |   |<TensorType(float64, vector)> [@E] <TensorType(float64, vector)>
 |TensorConstant{-1} [@F] <TensorType(int8, scalar)>
 |MakeVector [@G] <TensorType(int64, vector)> ''   
   |Shape_i{0} [@D] <TensorType(int64, scalar)> ''   
   |Shape_i{0} [@D] <TensorType(int64, scalar)> ''   
Split{2}.1 [@A] <TensorType(float64, vector)> ''   

Storage map footprint:
 - MakeVector.0, Shape: (2,), ElemSize: 8 Byte(s), TotalSize: 16 Byte(s)
 - TensorConstant{(1,) of 1.0}, Shape: (1,), ElemSize: 8 Byte(s), TotalSize: 8 Byte(s)
 - TensorConstant{-1}, Shape: (1,), ElemSize: 1 Byte(s), TotalSize: 1.0 Byte(s)
 - <TensorType(float64, vector)>, Shape: (10,), ElemSize: 8 Byte(s), TotalSize: 80 Byte(s)
 - Alloc.0, Shape: (10,), ElemSize: 8 Byte(s), TotalSize: 80 Byte(s)
```

I'm on a mac, using the anaconda python stack.

### PR DESCRIPTION
**Pull request by orhanf**
fix #2747

### CODE REVIEW

**Comment by nouiz** [comment]
Add as a test the code from the issue in the T_Join_and_Split class in test_basic.py

Also do the fix in perform and c_code in case axis is symbolic. At the same time, check that the value of the axis is a good value. (and add a test to make sure the code raise an error)

File: theano/tensor/basic.py [hunk 1]
```diff
@@ -3404,6 +3404,31 @@ def make_node(self, *axis_and_tensors):
 
         """
         axis, tensors = axis_and_tensors[0], axis_and_tensors[1:]
+
+        # Fix the axis if it is negative, since it crashes with fast_run
```
Comment (bouthilx) [inline theano/tensor/basic.py]: I believe we should move this to _make_node_internal. This would simplify GpuJoin code at the same time.

**Comment by orhanf** [comment]
@nouiz i've added the tests, also what you say about moving it to `_make_node_internal`

**Comment by bouthilx** [comment]
I made a PR to fix GpuJoin and it fixed Join at the same time (#2871). @nouiz was right, both tickets are tightly related, I now believe we should address both at the same time. 

@nouiz You might not agree with my use of assert_op. Can you take a look at the PR? I would like to know your opinion.

**Comment by abergeron** [comment]
This was fixed with #2871

**Pull request closed without merging**